### ISSUE TITLE
1. Implement `SOP.Generic (GHC.Generically a)` and 2. export SOP.Generically 

### ISSUE THREAD

**Issue by Icelandjack** [well-typed/generics-sop#153]
Here are 2 suggestions once we get `GHC.Generically` and `GHC.Generically1` in [base 4.17](https://github.com/ghc/ghc/blob/master/libraries/base/changelog.md).

Those are simply newtypes, if you have a definition in terms of `GHC.Generic` or `GHC.Generic1` then you give it an instance.

```haskell
-- GHC.Generics
newtype Generically    a = Generically  a
newtype Generically1 f a = Generically1 (f a)
```

1. `SOP.Generic` has an implementation in terms of  `GHC.Generic`, thus I propose the following instance:

    ```haskell
    instance (GHC.Generic a, GFrom a, GTo a, Rep a ~ SOP I (GCode a))
          => SOP.Generic (GHC.Generically a) where
      type Code (GHC.Generically a) = GCode a
      from = gfrom
      to   = gto
    ```

   This way an instance of `SOP.Generic T` can be derived: `deriving SOP.Generic via GHC.Generically T`.

2. I also propose defining a newtype with the same name as `Generically`, to be imported qualified with `SOP.Generically`, it will serve the same purpose but for `SOP.Generic` definitions. I think every data-type generic library should export a name like that. If I can implement `Binary` in terms of `SOP.Generic` I define `instance (SOP.Generic a, ..) => Binary (SOP.Generically a)`:

    ```haskell
    -- Generics.SOP
    newtype Generically a = Generically a
    ```
    
A newtype like `SOP.Generically` can be combined with a library like [generic-override](https://hackage.haskell.org/package/generic-override) that overrides the generic representation of an instance, so we decouple a generic implementation from modifying the generic implementation: 
 
 ```haskell
data X = X { s :: Int, p :: Int }
  deriving stock GHC.Generic
  
  -- 1.
  deriving SOP.Generic via GHC.Generically T
  
  -- 2. 
  deriving Monoid via SOP.Generically
    (Override X
      '[ "x" `As` Sum Int
       , "y" `As` Product Int
       ])
```

**Comment by kosmikus** [comment]
Thanks! In principle, I'm in support of both, although regarding part 2, I'm somewhat afraid that the presence of the type will go mostly unnoticed, because the library itself has hardly any generic functions, so it would basically be a type without instances.

**Comment by Icelandjack** [comment]
I think `Generically` should be defined where [`Generics` is defined](https://github.com/well-typed/generics-sop/blob/39a3badb10e11efc4f907db557f998412ffc1b2a/generics-sop/src/Generics/SOP/Universe.hs#L97), then other libraries can start attaching instances to it. Or at least, if you depend on *generics-sop* then you can create an instance of `Generically`.

What name should be chosen? At first I thought about naming them all `Generically` and importing it qualified `SOP.Generically` but I don't like the idea of dealing with the clash. Would you want `SOPGenerically` or something similar?

```haskell
import Generics.SOP hiding (Generically(..))
import qualified Generics.SOP as SOP (Generically)
```

**Comment by phadej** [comment]
There is already a name clash in `Generic` type-class name.

**Comment by Icelandjack** [comment]
So you're fine with exporting this as `Generically`?

**Comment by phadej** [comment]
> So you're fine with exporting this as Generically?

It's up to @kosmikus

**Comment by kosmikus** [comment]
Yes, as long as we don't rename the `Generic` class, I don't see an issue with calling the `newtype` `Generically` as well.

@Icelandjack do you want to submit a PR? Otherwise, I'll try to do this whenever I next get around to making changes on the library.

**Comment by Icelandjack** [comment]
I made a PR #159 and updated some of the documentation. I didn't build the project so there may be some errors.
